src/box/xrow.c

```diff
@@ -1944,6 +1962,57 @@ xrow_encode_vclock(struct xrow_header *row, const struct vclock *vclock)
 	return 0;
 }
 
+int
+xrow_decode_heartbeat(const struct xrow_header *row, struct vclock *vclock,
+		      uint64_t *vclock_sync)
+{
+	if (vclock != NULL)
+		vclock_create(vclock);
+	if (vclock_sync != NULL)
+		*vclock_sync = 0;
+	if (row->bodycnt == 0)
+		return 0;
+	const char *d = (const char *)row->body[0].iov_base;
+	if (mp_typeof(*d) != MP_MAP) {
+		xrow_on_decode_err(row, ER_INVALID_MSGPACK, "request body");
+		return -1;
+	}
+	uint32_t map_size = mp_decode_map(&d);
+	for (uint32_t i = 0; i < map_size; i++) {
+		if (mp_typeof(*d) != MP_UINT) {
+			mp_next(&d);
+			mp_next(&d);
+			continue;
+		}
+		unsigned key = mp_decode_uint(&d);
+		switch (key) {
+		case IPROTO_VCLOCK:
+			if (vclock == NULL)
+				goto skip;
+			if (mp_decode_vclock_ignore0(&d, vclock) != 0) {
+				xrow_on_decode_err(row, ER_INVALID_MSGPACK,
+						   "invalid vclock");
+				return -1;
+			}
+			break;
+		case IPROTO_VCLOCK_SYNC:
```

Comment:
This function looks like a duplicate of `xrow_decode_subscribe`, but with fewer keys. Maybe, instead of adding a new parser, add `IPROTO_VCLOCK_SYNC` to `xrow_decode_subscribe` and rename the latter to something like `xrow_decode_replication`, then make `xrow_decode_heartbeat` call it?

Reply:
I'm really concerned with `xrow_decode_heartbeat` having to switch-case over all the 'subscribe' keys. Like `IPROTO_INSTANCE_UUID`, `IPROTO_ID_FILTER` and so on.

I think it would be more or less fine for rare requests, like `subscribe` or `join`, but I don't want to trigger the same machinery for heartbeats, which will be sent quite often once linearizable read is introduced.

Reply:
Ok.

Reply:
FYI: in my future patch I am reworking all these replication request codecs to avoid abusing `xrow_decode_subscribe()` with tons of optional parameters for everything.  Don't necessarily need to dwell on it in this patch.